### PR TITLE
OH MY GOD -- I HAVENT COMMITED AT ALL

### PR DESCRIPTION
**Pull request by nerdbetter**
sooooooooo the findByIdAnd${something} isn't working like it shows in the docs..

### CODE REVIEW

File: jason-lab/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+## Chat API User DataBase
```
Comment (dahlbyk) [inline jason-lab/README.md]: YAY A README

File: jason-lab/model/chat-rooms.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+const mongoose = require('mongoose');
+const { Schema } = mongoose;
+const User = require('./user');
+const debug = require('debug')('app:chat-room');
+
+const roomSchema = Schema({
+  roomName: {type: String},
```
Comment (dahlbyk) [inline jason-lab/model/chat-rooms.js:9]: This should probably be `required`

File: jason-lab/lib/error-middleware.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+module.exports = function (err, req, res, next) {
```
Comment (dahlbyk) [inline jason-lab/lib/error-middleware.js]: `eslint` doesn't like that this `next` parameter is unused; however, if you remove it, Express can't tell the difference between `function (err, req, res)` and `function (req, res, next)`. So we have to keep the unused parameter:

```javascript
// eslint-disable-next-line no-unused-vars
module.exports = function (err, req, res, next) {
```

File: jason-lab/route/room-route.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+const router = module.exports = new Router();
+
+router.post('/api/room', jsonParser, function(req, res, next){
+  debug('POST: /api/room');
+  new Room(req.body).save()
+    .then(room => res.json(room))
+    .catch(err => next(err));
+});
+router.get('/api/room/:id', jsonParser, function(req, res, next){
```
Comment (dahlbyk) [inline jason-lab/route/room-route.js:15]: `jsonParser` isn't necessary for a `GET`

File: jason-lab/server.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+const express = require('express');
+const morgan = require('morgan');
+const mongoose = require('mongoose');
+const cors = require('cors');
+const debug = require('debug');
```
Comment (dahlbyk) [inline jason-lab/server.js]: Take a look at this...

File: jason-lab/route/user-route.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+router.post('/api/user', jsonParser, function(req, res, next){
+  debug('POST: /api/user');
+  new User(req.body).save()
+    .then(user => res.json(user))
+    .catch(err => next(err));
+});
+router.get('/api/user/:id', jsonParser, function(req, res, next){
+  debug('GET: /api/user');
+  User.findById(req.params.id)
+    .populate('users')
```
Comment (dahlbyk) [inline jason-lab/route/user-route.js:18]: `populate('users')` on your `User` doesn't seem right?

File: jason-lab/test/user-route.test.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+const request = require('supertest')(app);
+const {expect} = require('chai');
+
+const User = require('../model/user');
+
+describe('userroutes', function (){
+  describe('POST /api/user', function (){
+    describe('with a body', function (){
+      after(function (){
+        return User.remove(this.testUser);
```
Comment (dahlbyk) [inline jason-lab/test/user-route.test.js:11]: `this.testUser` isn't set in this context

File: jason-lab/test/room-route.test.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+const {expect} = require('chai');
+
+const Room = require('../model/chat-rooms');
+const User = require('../model/user');
+
+describe('roomroutes', function (){
+  describe('POST /api/room', function (){
+    describe('with a body', function (){
+      after(function (){
+        return Room.remove(this.testRoom);
```
Comment (dahlbyk) [inline jason-lab/test/room-route.test.js:12]: `this.testRoom` isn't set in this context

File: jason-lab/test/room-route.test.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+          expect(res.body.roomName).to.equal('updated');
+        });
+    });
+    it('should return 404 if missing id', function(){
+      return request
+        .put('/api/note/missing')
+        .expect(404);
+    });
+  });
+  describe('PUT with a valid id containing notes', function () {
```
Comment (dahlbyk) [inline jason-lab/test/room-route.test.js:58]: These are `GET` tests!

File: jason-lab/test/room-route.test.js [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            .then(user => this.testUser = user);
+        });
+    });
+    after(function () {
+      return Promise.all([
+        Room.remove({}),
+        User.remove({}),
+      ]);
+    });
+    it('should return a list', function () {
```
Comment (dahlbyk) [inline jason-lab/test/room-route.test.js:74]: a list of what?

File: jason-lab/test/room-route.test.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        .send({roomName:'updated'})
+        .expect(200)
+        .expect(res=>{
+          expect(res.body._id).to.equal(this.putRoom._id.toString());
+          expect(res.body.roomName).to.equal('updated');
+        });
+    });
+    it('should return 404 if missing id', function(){
+      return request
+        .put('/api/note/missing')
```
Comment (dahlbyk) [inline jason-lab/test/room-route.test.js:54]: Note?

File: jason-lab/test/user-route.test.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        .send({nickName:'updated'})
+        .expect(200)
+        .expect(res=>{
+          expect(res.body._id).to.equal(this.putUser._id.toString());
+          expect(res.body.nickName).to.equal('updated');
+        });
+    });
+    it('should return 404 if missing id', function(){
+      return request
+        .put('/api/note/missing')
```
Comment (dahlbyk) [inline jason-lab/test/user-route.test.js]: Note?

**Review by dahlbyk — COMMENTED** [review]
* [ ] `lint` passes
* [x] `test` passes

### Day 13

#### Feature Tasks
* [x] create an HTTP Server using `express`
* [x] create a resource **model** of your choice that uses `mongoose.Schema` and `mongoose.model`
* [x] use the `body-parser` express middleware to parse the `req` body on `POST` and `PUT` requests
* [x] use the npm `debug` module to log the functions and methods that are being used in your application
* [x] use the express `Router` to create a route for doing **RESTFUL CRUD** operations against your _model_

#### Server Endpoints
##### `/api/resource-name`
* [x] `POST` request
  * [x] should pass data as stringifed JSON in the body of a post request to create a new resource

##### `/api/resource-name/:id`
* `GET` request
  * [x] should pass the id of a resource through the url endpoint to get a resource
  * [x] **this should use `req.params`, not querystring parameters**
* `PUT` request
  * [x] should pass data as stringifed JSON in the body of a put request to update a pre-existing resource
  * [x] **this should use `req.params`**
* `DELETE` request
  * [x] should pass the id of a resource though the url endpoint to delete a resource
  * [x] **this should use `req.params`**

#### Tests
* [ ] create a test that will ensure that your API returns a status code of 404 for routes that have not been registered
* create a series of tests to ensure that your `/api/resource-name` endpoint responds as described for each condition below:
  * [x] `GET` - test 200, returns a resource with a valid body
  * [x] `GET` - test 404, respond with 'not found' for valid requests made with an id that was not found
  * [x] `PUT` - test 200, returns a resource with an updated body
  * [ ] `PUT` - test 400, responds with 'bad request' if no request body was provided
  * [ ] `PUT` - test 404, responds with 'not found' for valid requests made with an id that was not found
  * [x] `POST` - test 400, responds with 'bad request' if no request body was provided
  * [x] `POST` - test 200, returns a resource for requests made with a valid body

#### Bonus
* [ ] **2pts:** a `GET` request to `/api/resource-name` should return an array of stored resources

### Day 14

#### Feature Tasks
* [x] include an additional resource that contains a "relationship" to the single resource that has already been created
* create routes for your newly added resource
  * [x] `POST`
  * [x] `GET`
  * [x] `PUT`
  * [x] `DELETE`
* test your application to ensure that it meets the standard criteria of a working **full CRUD** REST API
  * [x] `POST`
  * [x] `GET`
  * [x] `PUT`
  * [x] `DELETE`
* [x] use `populate` in the `get()` route logic your  `/api/new-resource-name/:id` route to populate the associated property used to house related resources **(ex: `List.findById(req.params.id).populate('notes')`)**

File: jason-lab/README.md (hunk 1 above)
Comment (dahlbyk) [reply jason-lab/README.md]: Was so excited to see a readme that I forgot to look at it. :wink:

1. `/api/chat` doesn't exist now
2. Pay attention to which paths support which methods; as written, it's not clear which methods are supported with and without `/:id`

File: jason-lab/test/room-route.test.js (hunk 11 above)
Comment (nerdbetter) [reply jason-lab/test/room-route.test.js:54]: changed

File: jason-lab/test/user-route.test.js (hunk 7 above)
Comment (nerdbetter) [reply jason-lab/test/user-route.test.js:11]: I"M GONNA DELETE IT ANYWAY

**Pull request closed without merging**